### PR TITLE
Lucky chest

### PR DESCRIPTION
**Pull request by Steanky**

### CODE REVIEW

File: core/src/main/java/org/phantazm/core/player/BasicPlayerView.java [hunk 1]
```diff
@@ -126,6 +126,17 @@ private CompletableFuture<String> getUsernameRequest() {
...
+    public @NotNull Component getDisplayNameIfPresent() {
+        Optional<? extends Component> cached = getPlayer().map(Player::getDisplayName);
+        if (cached.isPresent()) {
+            return cached.get();
+        }
+
+        return getUsernameIfCached().map(Component::text).orElseGet(() -> Component.text(getUUID().toString()));
+
+    }
+
```
Comment (tahmid-23) [inline core/src/main/java/org/phantazm/core/player/BasicPlayerView.java:139]: very similar to the method right below (which I can't highlight)

Comment (Steanky) [reply core/src/main/java/org/phantazm/core/player/BasicPlayerView.java:139]: difference is it won't make API calls to mojang, it's basically getDisplayNameIfCached() (should probably call it that)

File: zombies/src/main/java/org/phantazm/zombies/map/shop/display/FlagConditionalDisplay.java [hunk 2]
```diff
@@ -33,20 +37,22 @@ public FlagConditionalDisplay(@NotNull Data data, @NotNull @Child("success") Lis
...
+        List<ShopDisplay> current = switch (data.source) {
+            case MAP -> flags.hasFlag(data.flag) ? success : failure;
+            case SHOP -> shop.flags().hasFlag(data.flag) ? success : failure;
+        };
 
         for (ShopDisplay display : current) {
             display.initialize(shop);
         }
+
+        this.current = current;
```
Comment (tahmid-23) [inline zombies/src/main/java/org/phantazm/zombies/map/shop/display/FlagConditionalDisplay.java:50]: why do you need a separate variable and field

File: zombies/src/main/java/org/phantazm/zombies/map/shop/FlagSource.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package org.phantazm.zombies.map.shop;
+
+public enum FlagSource {
```
Comment (tahmid-23) [inline zombies/src/main/java/org/phantazm/zombies/map/shop/FlagSource.java:3]: do we reasonably see this being used in places other than shops?

Comment (Steanky) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/FlagSource.java:3]: possibly, for now it's only used in shops

Comment (tahmid-23) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/FlagSource.java:3]: well I would make it ShopFlagSource or something to clarify
also if were to expand this enum then the switch statement might not make sense and might have to throw exc

File: core/src/main/java/org/phantazm/core/player/BasicPlayerView.java (hunk 1 above)
Comment (Steanky) [reply core/src/main/java/org/phantazm/core/player/BasicPlayerView.java:139]: well the other is an Optional which is a bit messier to work with

File: zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/ChangeSelectionGroupInteractor.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        if (data.excludeCurrent) {
+            if (interactors.size() == 1) {
+                return true;
+            }
+
+            int newIndex = random.nextInt(interactors.size() - 1);
+            if (newIndex >= currentActiveInteractorIndex) {
+                newIndex++;
+            }
```
[the remark this answers is not in the document]
Comment (Steanky) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/ChangeSelectionGroupInteractor.java:64]: newIndex can't equal interactors.size() - 1, random#nextInt is exclusive

File: core/src/main/java/org/phantazm/core/player/BasicPlayerView.java (hunk 1 above)
Comment (tahmid-23) [reply core/src/main/java/org/phantazm/core/player/BasicPlayerView.java:139]: w/e its not that important

File: zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/PlaySongInteractor.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        if (!songLoader.songs().contains(data.songKey)) {
+            return false;
+        }
+
+        List<SongPlayer.Note> notes = songLoader.getNotes(data.songKey);
+        if (data.broadcast) {
+            interaction.player().getPlayer().ifPresent(player -> {
+                Instance instance = player.getInstance();
+                if (instance != null) {
+                    songPlayer.play(instance, Sound.Emitter.self(), notes);
```
[the remark this answers is not in the document]
Comment (Steanky) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/PlaySongInteractor.java:45]: instance is a forwarding audience, so the way it should work is the sound is played as Sound.Emitter.self() to everyone

[the remark this answers is not in the document]
Comment (tahmid-23) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/PlaySongInteractor.java:45]: yeah but I shouldn't hear a chest on the other end of the map

File: zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+        updateHologram(newComponents);
+    }
+
+    private TagResolver[] getTagsForFrame(SlotMachineFrame frame, TagResolver... additionalTags) {
+        TagResolver rollingPlayerTag = Placeholder.component("rolling_player",
+                rollInteraction.player().module().getPlayerView().getDisplayNameIfPresent());
+
+        Component displayName = frame.getVisual().getDisplayName();
+        if (displayName == null) {
+            displayName = Component.text(frame.getVisual().material().key().value());
```
Comment (tahmid-23) [inline zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java:217]:
```suggestion
            displayName = Component.translatable(frame.getVisual().material().registry().translationKey());
```

File: zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            Component newComponent = newComponents.get(i);
+            Component oldComponent = i < hologram.size() ? hologram.get(i) : null;
+
+            if (!newComponent.equals(oldComponent)) {
+                hologram.set(i, newComponent);
+            }
+        }
+    }
+
+    public interface SlotMachineFrame {
```
Comment (tahmid-23) [inline zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java:245]: why do we need need an interface here for a primarily internal class?

Comment (Steanky) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java:245]: just to make it easier to expand ig

Comment (tahmid-23) [reply zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java:245]: it is mostly personal preference but I think it just needlessly adds complexity

File: zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+            }
+        }
+
+        if (currentFrameIndex == data.frameCount) {
+            if (!doneRolling) {
+                rollFinishTime = time;
+            }
+
+            doneRolling = true;
+            long ticksSinceDoneRolling = time - rollFinishTime / MinecraftServer.TICK_MS;
```
Comment (tahmid-23) [inline zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java:148]: do you need () here

**Review by tahmid-23 — APPROVED** [review]
now make PiglinBarterInteractor

**Pull request merged**